src/v/kafka/protocol/schemata/generator.py

```diff
@@ -468,7 +471,29 @@ def guard(self):
             if self.max != None:
                 cond.append(f"version <= api_version({self.max})")
             cond = " && ".join(cond)
-        return cond
+
+        return (self.guard_enum.NO_GUARD,
+                None) if cond == "" else (self.guard_enum.GUARD, cond)
+
+    def guard(self, flex, first_flex):
+        """
+        Optimize generated code by either omitting a guard or source itself
+        """
+        if first_flex < 0:
+            return self._guard()
+        elif not flex:
+            if self.min >= first_flex:
+                return self.guard_enum.NO_SOURCE, None
+        else:
+            if self.max == None:
+                if self.min <= first_flex:
+                    return self.guard_enum.NO_GUARD, None
+            elif self.max < first_flex:
+                return self.guard_enum.NO_SOURCE, None
```

Comment:
i don't really understand what NO_SOURCE is

Reply:
Yeah, I myself preferred how this looked before too, but to explain NO_SOURCE means actually don't generate any source at all - conditional and condition body too. For example lets say `flexibleVersions: 7+` for a request, and the generator macro is generating `encode_pre_flex`. In cases where a field has a version >= 7, the `version_check` macro will just not generate any source, instead of doing so and wrapping in a conditional that is guaranteed to never be true.